cse/indexing/DocumentMapBuilder.py

```diff
@@ -0,0 +1,105 @@
+import os
+import errno
+import shutil
+
+from cse.util import PackerUtil
+from cse.indexing.commons import PARTIAL_THRESHOLD, SKIP_SIZE
+
+
+class DocumentMapBuilder:
+
+    def __init__(self, document_map_index, document_map_dict):
+        if not os.path.exists(os.path.dirname(document_map_index)):
+            try:
+                os.makedirs(os.path.dirname(document_map_index))
+            except OSError as exc: # Guard against race condition
+                if exc.errno != errno.EEXIST:
+                    raise
+
+        self.__document_map_index = document_map_index
+        self.__document_map_dict = document_map_dict
+        self.__index = []
+        self.__i = 0
+        self.__content = 0
+        if os.path.exists(self.__document_map_index):
+            os.remove(self.__document_map_index)
+        if os.path.exists(self.__document_map_dict):
+            os.remove(self.__document_map_dict)
+        dir = os.path.dirname(self.__document_map_index)
+        if os.path.isdir(os.path.join(dir, 'document_map_partials')):
+            shutil.rmtree(os.path.join(dir, 'document_map_partials'))
+
+    def close(self):
+        self.__index.sort()
+        self.__save_partial_index(self.__index, self.__i)
+        doc_dict = self._build_index_and_dict()
+        PackerUtil.packToFile(doc_dict, self.__document_map_dict, type=PackerUtil.PICKLE)
```

Comment:
please also clear all internal state here, like you do in line 42 to 46

Reply:
done